File: objc/Tests/TestSpectralDatasetImages.m
/*
 * TestSpectralDatasetImages.m — P2.5 OIT2.
 *
 * Exercises the uniform image-collection accessors that replace the
 * three typed properties (-msImage / -ramanImage / -irImage) on
 * TTIOSpectralDataset:
 *
 *   - (nullable TTIOImage *)imageForKind:(TTIOImageKind)kind;
 *   - (NSArray<TTIOImage *> *)images;
 *
 * Consumers needing the concrete subclass cast the polymorphic result
 * (e.g. (TTIOMSImage *)[ds imageForKind:TTIOImageKindMS]). The lazy
 * file-backed materialisation and per-kind caching are preserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"

#import "Dataset/TTIOSpectralDataset.h"
#import "Image/TTIOImage.h"
#import "Image/TTIOMSImage.h"
#import "Image/TTIORamanImage.h"
#import "Image/TTIOIRImage.h"
#include <unistd.h>

static NSString *makeImagesTempPath(NSString *suffix)
{
    NSString *base = [NSString stringWithFormat:@"ttio_dsimages_%d_%@.tio",
                       (int)getpid(), suffix];
    return [NSTemporaryDirectory() stringByAppendingPathComponent:base];
}

static TTIOMSImage *buildMSFixture(void)
{
    const NSUInteger W = 2, H = 2, SP = 3;
    NSMutableData *cube = [NSMutableData dataWithLength:W * H * SP * sizeof(double)];
    NSMutableData *mz = [NSMutableData dataWithLength:SP * sizeof(double)];
    double *mp = (double *)mz.mutableBytes;
    for (NSUInteger i = 0; i < SP; i++) mp[i] = 100.0 * (i + 1);
    return [[TTIOMSImage alloc] initWithTitle:@"images_fixture"
                           isaInvestigationId:@""
                              identifications:@[]
                              quantifications:@[]
                            provenanceRecords:@[]
                                        width:W
                                       height:H
                               spectralPoints:SP
                                     tileSize:0
                                   pixelSizeX:1.0
                                   pixelSizeY:1.0
                                  scanPattern:@"raster"
                                         cube:cube
                                       mzAxis:mz];
}

// -- 1. imageForKind: returns the MS image and nil for absent kinds --------

static void testImageForKindMSPresentRamanNil(void)
{
    TTIOMSImage *img = buildMSFixture();
    NSString *path = makeImagesTempPath(@"forkind");
    unlink([path fileSystemRepresentation]);
    NSError *err = nil;
    PASS([img writeToFilePath:path error:&err] && err == nil,
         "imageForKind: wrote MS image .tio");

    TTIOSpectralDataset *ds = [TTIOSpectralDataset readFromFilePath:path
                                                                error:&err];
    PASS(ds != nil && err == nil, "imageForKind: opened dataset");

    TTIOImage *result = [ds imageForKind:TTIOImageKindMS];
    PASS(result != nil, "imageForKind:MS returns non-nil for MS-bearing dataset");
    PASS([result isKindOfClass:[TTIOMSImage class]],
         "imageForKind:MS returns a TTIOMSImage instance");

    PASS([ds imageForKind:TTIOImageKindRaman] == nil,
         "imageForKind:Raman is nil when no raman_image_cube present");

    [ds closeFile];
    unlink([path fileSystemRepresentation]);
}

// -- 2. -images contains only the present kinds ----------------------------

static void testImagesCollectionPresentOnly(void)
{
    TTIOMSImage *img = buildMSFixture();
    NSString *path = makeImagesTempPath(@"images");
    unlink([path fileSystemRepresentation]);
    NSError *err = nil;
    PASS([img writeToFilePath:path error:&err] && err == nil,
         "images: wrote MS image .tio");

    TTIOSpectralDataset *ds = [TTIOSpectralDataset readFromFilePath:path
                                                                error:&err];
    PASS(ds != nil && err == nil, "images: opened dataset");

    NSArray<TTIOImage *> *images = [ds images];
    PASS(images.count == 1,
         "images contains exactly the one present (MS) image");
    PASS(images.count == 1
         && [images[0] isKindOfClass:[TTIOMSImage class]],
         "images[0] is the MS image");

    [ds closeFile];
    unlink([path fileSystemRepresentation]);
}

void testSpectralDatasetImages(void);
void testSpectralDatasetImages(void)
{
    testImageForKindMSPresentRamanNil();
    testImagesCollectionPresentOnly();
}

File: python/tests/test_spectral_dataset_images.py
"""IPT2: SpectralDataset's uniform image collection.

The three typed lazy accessors (``image`` / ``raman_image`` /
``ir_image``) are replaced by a uniform :meth:`image_for_kind` lookup
and an :attr:`images` collection keyed by :class:`ttio.enums.ImageKind`.
The lazy per-kind read+cache behaviour is preserved; the old accessors
are removed.
"""
from __future__ import annotations

from pathlib import Path

import numpy as np
import pytest

from ttio import MSImage, SpectralDataset
from ttio.enums import ImageKind


def _build_image(w: int, h: int, sp: int) -> MSImage:
    cube = np.arange(h * w * sp, dtype=np.float64).reshape(h, w, sp) * 0.1
    mz = np.linspace(100.0, 100.0 + (sp - 1) * 100.0, sp)
    return MSImage(
        width=w, height=h, spectral_points=sp,
        intensity=cube,
        mz_axis=mz,
        pixel_size_x=10.0, pixel_size_y=10.0, scan_pattern="raster",
    )


def _write_ms_only(out: Path) -> None:
    img = _build_image(2, 2, 4)
    SpectralDataset.write_minimal(
        out, title="img-test", isa_investigation_id="",
        runs={}, image=img,
    )


def test_image_for_kind_returns_present_kind(tmp_path: Path) -> None:
    out = tmp_path / "ms_only.tio"
    _write_ms_only(out)
    with SpectralDataset.open(out) as ds:
        ms = ds.image_for_kind(ImageKind.MS)
        assert ms is not None
        assert isinstance(ms, MSImage)
        assert ms.width == 2


def test_image_for_kind_absent_kind_is_none(tmp_path: Path) -> None:
    out = tmp_path / "ms_only.tio"
    _write_ms_only(out)
    with SpectralDataset.open(out) as ds:
        assert ds.image_for_kind(ImageKind.RAMAN) is None
        assert ds.image_for_kind(ImageKind.IR) is None


def test_images_collection_contains_only_present_kinds(tmp_path: Path) -> None:
    out = tmp_path / "ms_only.tio"
    _write_ms_only(out)
    with SpectralDataset.open(out) as ds:
        images = ds.images
        assert isinstance(images, dict)
        assert set(images.keys()) == {ImageKind.MS}
        assert isinstance(images[ImageKind.MS], MSImage)


def test_image_for_kind_caches(tmp_path: Path) -> None:
    """Repeated image_for_kind calls return the same cached object."""
    out = tmp_path / "ms_only.tio"
    _write_ms_only(out)
    with SpectralDataset.open(out) as ds:
        first = ds.image_for_kind(ImageKind.MS)
        second = ds.image_for_kind(ImageKind.MS)
        assert first is second, "second call should return cached object"


def test_image_for_kind_rejects_unknown_kind(tmp_path: Path) -> None:
    out = tmp_path / "ms_only.tio"
    _write_ms_only(out)
    with SpectralDataset.open(out) as ds:
        with pytest.raises(ValueError, match="unknown ImageKind"):
            ds.image_for_kind(object())  # type: ignore[arg-type]


def test_old_accessors_removed(tmp_path: Path) -> None:
    """The typed image/raman_image/ir_image properties are gone."""
    out = tmp_path / "ms_only.tio"
    _write_ms_only(out)
    with SpectralDataset.open(out) as ds:
        for attr in ("image", "raman_image", "ir_image"):
            assert not hasattr(ds, attr), (
                f"old accessor {attr!r} must be removed"
            )
